gc/2021/GC-07-27.md

```diff
@@ -0,0 +1,481 @@
+![WebAssembly logo](/images/WebAssembly.png)
+
+## Agenda for the July 27 video call of WebAssembly's Garbage Collection Subgroup
+
+- **Where**: zoom.us
+- **When**: July 27, 4pm-6pm UTC (July 13, 9am-11am Pacific Daylight Time)
+- **Location**: *link on calendar invite*
+
+### Registration
+
+None required if you've attended before. Fill out the form here to sign up if
+it's your first time: https://forms.gle/JehrAB4gWbtHjybt9. The meeting is open
+to CG members only.
+
+## Logistics
+
+The meeting will be on a zoom.us video conference.
+Installation is required, see the calendar invite.
+
+## Agenda items
+
+1. Opening, welcome and roll call
+    1. Opening of the meeting
+    1. Introduction of attendees
+1. Find volunteers for note taking (acting chair to volunteer)
+1. Adoption of the agenda
+1. Proposals and discussions
+    1. Discussion: Nominal and structural types (2 hours)
+1. Closure
+
+## Meeting Notes
+
+### Introduction of attendees
+
+- Thomas Lively
+- Alon Zakai
+- Tim Steenvoorden
+- Jay Phelps
+- Igor Laevsky
+- Daniel Wirtz
+- Ross Tate
+- Ryan Hunt
+- Zalim Bashorov
+- Dmitry Bezhetskov
+- Sergey Rubanov
+- Andreas Rossberg
+- Conrad Watt
+- Jakob Kummerow
+- Ioanna Dimitriou
+- Francis McCabe
+- Rick Battagline
+- Emanuel Ziegler
+- Asumu Takikawa
+- Adam Klein
+- Luke Wagner
+- Slava Kuzmich
+- Daniel Ehrenberg
+- Ben Titzer
+- Keith Miller
+- Petr Penzin
+- Zhi An Ng
+
+### Discussion: Nominal and structural types (2 hours)
+
+TODO: slides
+
+AR: Recap from last time: there are many degrees of modularity that are important for different use cases and all are possible to use in Wasm today. Anything that works in a modular fashion today should still work with GC. Structural types are not a fringe problem. Many languages have structural types and they must be compiled correctly and handled at module boundaries.
+
+Structural types are infinite in number, so declaring them in a runtime module does not scale.
+
+CW: I appreciate the number of possible tuple types is infinite. You’re still matching each tuple type in the source code to a different struct definition. Same number of definitions in nominal and structural.
+
+AR: Right. This is about the runtime module not scaling.
+
+RT: You would have a loader that would deduplicate these types.
+
+AR: Will get to that.
+
+RT: How do you allocate a new struct for this?
+
+AR: Just allocate a struct.
+
+RT: In what OO language does a tuple not have a vtable?
+
+AR: Good question. Not sure.
+
+BT: This seems like a distraction. It’s easy to come up with a language with tuples with no vtables.
+
+RT: Languages that do have tuples with vtables also need a mechanism to consolidate their vtables. If we have structural types, we can’t service such languages.
+
+AR: You can synthesize the vtable, accessors, etc. on the fly.
+
+RT: Java does ….
```

Comment:
@RossTate do you remember the rest of what you were saying here?

Reply:
Hmm, I think I was pointing out that having separate Java modules synthesize their own v-tables for each kind of array they construct is not a realistic solution both because of direct overheads (allocation time and space) and indirect overheads (many optimizations rely on every class having at most one interface).